File: 20-valid-parentheses/valid-parentheses.scala
import scala.collection.mutable

object Solution {
    def isValid(s: String): Boolean = {
        def iter(chars: Array[Char], stack: Array[Char]): Boolean =
            chars.headOption match
                case Some(char) =>
                    char match
                        case '{' | '[' | '(' => iter(chars.tail, stack :+ char)
                        case _               =>
                            val peek = stack.lastOption.getOrElse("")
                            if (peek == '{' && char == '}') ||
                               (peek == '[' && char == ']') ||
                               (peek == '(' && char == ')')
                            then iter(chars.tail, stack.init)
                            else false
                case None       => stack.isEmpty

        iter(s.toArray, Array.empty[Char])
    }
}

File: 20-valid-parentheses/valid-parentheses.java
import java.util.Stack;

class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();

        for(char c : s.toCharArray()){
            if(c == '{' || c == '[' || c == '(') {
                stack.push(c);
            } else if(stack.isEmpty()){
                return false;
            } else {
                char peek = stack.pop();
                if(c == '}' && peek != '{'){
                    return false;
                }
                if(c == ']' && peek != '['){
                    return false;
                }
                if(c == ')' && peek != '('){
                    return false;
                }
            }
        }
        
        return stack.isEmpty();
    }
}